### PR TITLE
Close #18317: Refactor InstallTrack window to class

### LINKED ISSUE #18317: Refactor window to class: InstallTrack
Refers to file: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2-ui/windows/InstallTrack.cpp
See #13737 for more details on how to accomplish the task.

### PR DESCRIPTION
**Pull request by jacknull1991**
Fixes #18317

Also I issued a bug report #18411 about this window.

### CODE REVIEW

File: src/openrct2-ui/windows/InstallTrack.cpp [hunk 1]
```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
...
-
-    screenPos = w->windowPos + ScreenCoordsXY{ widget->midX(), widget->bottom - 12 };
-
-    // Warnings
-    const TrackDesign* td6 = _trackDesign.get();
-    if (td6->track_flags & TRACK_DESIGN_FLAG_SCENERY_UNAVAILABLE)
+    void OnTextInput(WidgetIndex widgetIndex, std::string_view text) override
     {
-        if (!gTrackDesignSceneryToggle)
+        if (widgetIndex != WIDX_INSTALL || str_is_null_or_empty(std::string(text).c_str()))
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/InstallTrack.cpp:126]:
```suggestion
        if (widgetIndex != WIDX_INSTALL || text.empty())
```

File: src/openrct2-ui/windows/InstallTrack.cpp [hunk 2]
```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
...
-                dpi, screenPos, 308, STR_DESIGN_INCLUDES_SCENERY_WHICH_IS_UNAVAILABLE, {}, { TextAlignment::CENTRE });
-            screenPos.y -= LIST_ROW_HEIGHT;
+            return;
         }
-    }
 
-    // Information
-    screenPos = w->windowPos + ScreenCoordsXY{ widget->left + 1, widget->bottom + 4 };
-    // 0x006D3CF1 -- 0x006d3d71 missing
+        _trackName = static_cast<std::string>(text);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/InstallTrack.cpp:131]:
```suggestion
        _trackName = std::string(text);
```

File: src/openrct2-ui/windows/InstallTrack.cpp [hunk 3]
```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
...
+    }
+};
 
-static void WindowInstallTrackDesign(rct_window* w)
+rct_window* WindowInstallTrackOpen(const utf8* path)
 {
-    auto env = OpenRCT2::GetContext()->GetPlatformEnvironment();
-    auto destPath = env->GetDirectoryPath(OpenRCT2::DIRBASE::USER, OpenRCT2::DIRID::TRACK);
-    if (!Platform::EnsureDirectoryExists(destPath.c_str()))
+    std::unique_ptr<TrackDesign> trackDesign = TrackDesignImport(path);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/InstallTrack.cpp:405]:
```suggestion
    auto trackDesign = TrackDesignImport(path);
```

File: src/openrct2-ui/windows/InstallTrack.cpp [hunk 4]
```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
...
-        case WIDX_TOGGLE_SCENERY:
-            gTrackDesignSceneryToggle = !gTrackDesignSceneryToggle;
-            WindowInstallTrackUpdatePreview();
-            w->Invalidate();
-            break;
-        case WIDX_INSTALL:
-            WindowInstallTrackDesign(w);
-            break;
+        widgets = window_install_track_widgets;
+        track_list.track_list_being_updated = false;
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/InstallTrack.cpp:85]: This used to be after `WindowInitScrollWidgets` (not sure if it mattered)

File: src/openrct2-ui/windows/InstallTrack.cpp [hunk 5]
```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
...
+
+    gTrackDesignSceneryToggle = false;
+    _currentTrackPieceDirection = 2;
+
+    int32_t screenWidth = context_get_width();
+    int32_t screenHeight = context_get_height();
+    int32_t x = screenWidth / 2 - 201;
+    int32_t y = std::max(TOP_TOOLBAR_HEIGHT + 1, screenHeight / 2 - 200);
+
+    auto* window = WindowFocusOrCreate<InstallTrackWindow>(WindowClass::InstallTrack, ScreenCoordsXY(x, y), WW, WH, 0);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/InstallTrack.cpp:435]:
```suggestion
    auto screenPos = ScreenCoordsXY{ screenWidth / 2 - 201, std::max(TOP_TOOLBAR_HEIGHT + 1, screenHeight / 2 - 200) };

    auto* window = WindowFocusOrCreate<InstallTrackWindow>(WindowClass::InstallTrack, screenPos, WW, WH, 0);
```

File: src/openrct2-ui/windows/InstallTrack.cpp (hunk 4 above)
Comment (jacknull1991) [reply src/openrct2-ui/windows/InstallTrack.cpp:85]: From what I understand the order does not matter here.

Comment (tupaschoal) [reply src/openrct2-ui/windows/InstallTrack.cpp:85]: I think it doesn't either.

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**